scipy/optimize/optimize.py

```diff
@@ -490,10 +497,15 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
     if retall:
         allvecs = [sim[0]]
 
-    if maxiter is None:
+    if maxiter is None and maxfun is None:
         maxiter = N * 200
-    if maxfun is None:
         maxfun = N * 200
+    elif maxiter is not None and maxfun is not None:
+        pass
+    elif maxiter is not None and maxiter != np.inf:
+        maxfun = np.inf
+    elif maxfun is not None and maxfun != np.inf:
+        maxiter = np.inf
```

Comment:
`maxiter=None, maxfun=np.inf` is a tricky one. That will turn it into `maxiter=np.inf, maxfun=np.inf`, which might result in an infinite loop if there is some function that does not converge.

My (somewhat awkward) version was trying to avoid the case setting them both to inf, i.e. only set maxfun to inf if maxiter has a finite value.

Maybe it should interpret `maxiter=None, maxfun=np.inf`, as "never stop on maxfun, but use the default for maxiter", then it would become `maxiter=N*200, maxfun=np.inf`.

```

    # If neither are set, then set both to default
    if maxiter is None and maxfun is None:
        maxiter = N * 200
        maxfun = N * 200
    # Convert remaining Nones, to np.inf, unless the other is np.inf, in
    # which case use the default to avoid unbounded iteration
    elif maxiter is None:
        if maxfun is np.inf:
            maxiter = N * 200
        else:
            maxiter = np.inf
    elif maxfun is None:
        if maxiter is np.inf:
            maxfun = N * 200
        else:
            maxfun = np.inf

```

I am not 100% sure which is best.

If the users requests just maxfun=5000, then I think the code should respect that and not stop earlier (as it originally did). In terms of negative effects on existing code, this could make them take longer, but it would still definitely terminate.

If the user requests just maxfun=np.inf, should that be taken literally, or should the user be protected against infinite loops? The old code protected them. If they really want iteration to never stop without convergence they can set both maxfun and maxiter to np.inf.

(Either way, for the fixed version, should I 1) just add a new commit to this branch, 2) make a new branch and new pull request, 3) overwrite this branch with a push --force?).